source/common/upstream/cluster_factory_impl.h

```diff
@@ -0,0 +1,222 @@
+   */
+  static ClusterSharedPtr
+  create(const envoy::api::v2::Cluster& cluster, ClusterManager& cluster_manager,
+         Stats::Store& stats, ThreadLocal::Instance& tls,
+         Network::DnsResolverSharedPtr dns_resolver, Ssl::ContextManager& ssl_context_manager,
+         Runtime::Loader& runtime, Runtime::RandomGenerator& random, Event::Dispatcher& dispatcher,
+         AccessLog::AccessLogManager& log_manager, const LocalInfo::LocalInfo& local_info,
+         Server::Admin& admin, Singleton::Manager& singleton_manager,
+         Outlier::EventLoggerSharedPtr outlier_event_logger, bool added_via_api, Api::Api& api);
+
+  /**
+   * Create a dns resolver to be used by the cluster.
+   */
+  Network::DnsResolverSharedPtr selectDnsResolver(const envoy::api::v2::Cluster& cluster,
+                                                  ClusterFactoryContext& context);
+
+  // Upstream::ClusterFactory
+  ClusterSharedPtr create(const envoy::api::v2::Cluster& cluster,
+                          ClusterFactoryContext& context) override;
+  std::string name() override { return name_; }
+
+protected:
+  ClusterFactoryImplBase(const std::string& name) : name_(name) {}
+
+private:
+  /**
+   * Create an instance of ClusterImplBase.
+   */
+  virtual ClusterImplBaseSharedPtr
+  createClusterImpl(const envoy::api::v2::Cluster& cluster, ClusterFactoryContext& context,
+                    Server::Configuration::TransportSocketFactoryContext& socket_factory_context,
+                    Stats::ScopePtr&& stats_scope) PURE;
+  const std::string name_;
+};
+
+/**
+ * Factory for StaticClusterImpl cluster.
+ */
+class StaticClusterFactory : public ClusterFactoryImplBase {
+public:
+  StaticClusterFactory()
+      : ClusterFactoryImplBase(Extensions::Clusters::ClusterTypes::get().Static) {}
+
+private:
+  ClusterImplBaseSharedPtr
+  createClusterImpl(const envoy::api::v2::Cluster& cluster, ClusterFactoryContext& context,
+                    Server::Configuration::TransportSocketFactoryContext& socket_factory_context,
+                    Stats::ScopePtr&& stats_scope) override;
+};
+
+/**
+ * Factory for StrictDnsClusterImpl
+ */
+class StrictDnsClusterFactory : public ClusterFactoryImplBase {
+public:
+  StrictDnsClusterFactory()
+      : ClusterFactoryImplBase(Extensions::Clusters::ClusterTypes::get().StrictDns) {}
+
+private:
+  ClusterImplBaseSharedPtr
+  createClusterImpl(const envoy::api::v2::Cluster& cluster, ClusterFactoryContext& context,
+                    Server::Configuration::TransportSocketFactoryContext& socket_factory_context,
+                    Stats::ScopePtr&& stats_scope) override;
+};
+
+/**
+ * Common base class for custom cluster factory with custom configuration.
+ * @tparam ConfigProto is the configuration protobuf.
+ */
+template <class ConfigProto> class ConfigurableClusterFactoryBase : public ClusterFactoryImplBase {
+
+public:
+  /**
+   * @return ProtobufTypes::MessagePtr create empty config proto message for v2. The filter
```

Comment:
v1 is gone and this new extension only supports v2, so please update the comment accordingly.